aiida_quantumespresso/parsers/ph.py

```diff
@@ -16,44 +18,85 @@ class PhParser(Parser):
         filepath_dynmat = PhCalculation._FOLDER_DYNAMICAL_MATRIX
 
-        if filename_stdout not in list_of_files:
-            self.logger.error("The standard output file '{}' was not found but is required".format(filename_stdout))
-            return self.exit_codes.ERROR_READING_OUTPUT_FILE
+        if filename_stdout not in self.retrieved.list_object_names():
+            return self.exit_codes.ERROR_OUTPUT_STDOUT_MISSING
 
-        xml_tensor_file = None
-        if filename_tensor in list_of_files:
-            xml_tensor_file = output_folder._repository._get_base_folder().get_abs_path(filename_tensor)
+        try:
+            stdout = self.retrieved.get_object_content(filename_stdout)
+        except IOError:
+            return self.exit_codes.ERROR_OUTPUT_STDOUT_READ
+
+        try:
+            tensor_file = self.retrieved.get_object_content(filename_tensor)
+        except IOError:
+            tensor_file = None
 
         # Look for dynamical matrices
-        dynmat_dir = output_folder._repository._get_base_folder().get_abs_path(filepath_dynmat)
-        dynamical_matrix_list = [
-            os.path.join(dynmat_dir, file)
-            for file in os.listdir(dynmat_dir)
-            if os.path.isfile(os.path.join(dynmat_dir, file)) and
-            file.startswith(os.path.split(PhCalculation._OUTPUT_DYNAMICAL_MATRIX_PREFIX)[1]) and
-            not file.endswith('.freq')
-        ]
-
-        # Sort according to the number at the end of the dynamical matrix files, when there is one
+        dynmat_files = []
+        dynmat_folder = PhCalculation._FOLDER_DYNAMICAL_MATRIX
+        dynmat_prefix = os.path.split(PhCalculation._OUTPUT_DYNAMICAL_MATRIX_PREFIX)[1]
+
+        for filename in sorted(self.retrieved.list_object_names(dynmat_folder)):
+            if not filename.startswith(dynmat_prefix) or filename.endswith('.freq'):
+                continue
+
+            dynmat_files.append(self.retrieved.get_object_content(os.path.join(dynmat_folder, filename)))
+
         try:
-            dynamical_matrix_list = sorted(dynamical_matrix_list, key=lambda x: int(x.split('-')[-1]))
-        except ValueError:
-            dynamical_matrix_list = sorted(dynamical_matrix_list)
+            parsed_data, logs = parse_stdout(stdout, tensor_file, dynmat_files)
+        except Exception as exception:
+            self.logger.error(traceback.format_exc())
+            return self.exit_codes.ERROR_UNEXPECTED_PARSER_EXCEPTION
+
+        self.emit_logs(logs)
+        self.out('output_parameters', orm.Dict(dict=parsed_data))
+
+        if 'ERROR_OUT_OF_WALLTIME' in logs['error']:
+            return self.exit_codes.ERROR_OUT_OF_WALLTIME
+
+        if 'ERROR_CONVERGENCE_NOT_REACHED' in logs['error']:
+            return self.exit_codes.ERROR_CONVERGENCE_NOT_REACHED
+
+    def emit_logs(self, *args):
+        """Emit the messages in one or multiple "log dictionaries" through the logger of the parser.
```

Comment:
Since this method is the same in `PwParser` except for the ignored messages, does it make sense to factor it out as a utility function? We could keep these methods, but they would just pass the `ignore` list to the external function as an additional argument. If you want to do it later, that's fine.

Reply:
I kept it separate on purpose, but if it becomes useful for more than these two, then yeah we should refactor

Reply:
Ok.